Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,726 @@
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error()) {
+        return {};
+    }
+
+    return inode;
+}
+
+void ISO9660Inode::flush_metadata()
+{
+}
+
+KResultOr<size_t> ISO9660Inode::write_bytes(off_t, size_t, const UserOrKernelBuffer&, FileDescription*)
+{
+    return EROFS;
+}
+
+KResultOr<NonnullRefPtr<Inode>> ISO9660Inode::create_child(StringView, mode_t, dev_t, uid_t, gid_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::add_child(Inode&, const StringView&, mode_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::remove_child(const StringView&)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::chmod(mode_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::chown(uid_t, gid_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::truncate(u64)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_atime(time_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_ctime(time_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_mtime(time_t)
+{
+    return EROFS;
+}
+
+void ISO9660Inode::one_ref_left()
+{
+}
+
+ISO9660Inode::ISO9660Inode(ISO9660FS& fs, ISO::DirectoryRecordHeader const* record, StringView const& name)
+    : Inode(fs, get_inode_index(record, name))
+    , m_record(*record)
+{
+    dbgln_if(ISO9660_VERY_DEBUG, "Creating inode #{}", index());
+    create_metadata();
+}
+
+ISO9660Inode::~ISO9660Inode()
+{
+}
+
+KResultOr<NonnullRefPtr<ISO9660Inode>> ISO9660Inode::create_from_directory_record(ISO9660FS& fs, ISO::DirectoryRecordHeader const* record, StringView const& name)
+{
+    auto result = adopt_ref_if_nonnull(new (nothrow) ISO9660Inode(fs, record, name));
+    if (!result) {
+        return ENOMEM;
+    }
+    return result.release_nonnull();
+}
+
+void ISO9660Inode::create_metadata()
+{
+    u32 data_length = LittleEndian { m_record.data_length.little };
+    bool is_directory = has_flag(m_record.file_flags, ISO::FileFlags::Directory);
+    time_t recorded_at = parse_numerical_date_time(m_record.recording_date_and_time);
+
+    m_metadata = {
+        .inode = identifier(),
+        .size = data_length,
+        .mode = static_cast<mode_t>((is_directory ? S_IFDIR : S_IFREG) | (is_directory ? 0555 : 0444)),
+        .uid = 0,
+        .gid = 0,
+        .link_count = 1,
+        .atime = recorded_at,
+        .ctime = recorded_at,
+        .mtime = recorded_at,
+        .dtime = 0,
+        .block_count = 0,
+        .block_size = 0,
+        .major_device = 0,
+        .minor_device = 0,
+    };
+}
+
+time_t ISO9660Inode::parse_numerical_date_time(ISO::NumericalDateAndTime const& date)
+{
+    i32 year_offset = date.years_since_1900 - 70;
+
+    return (year_offset * 60 * 60 * 24 * 30 * 12)
+        + (date.month * 60 * 60 * 24 * 30)
+        + (date.day * 60 * 60 * 24)
+        + (date.hour * 60 * 60)
+        + (date.minute * 60)
+        + date.second;
+}
+
+StringView ISO9660Inode::get_normalized_filename(ISO::DirectoryRecordHeader const* record, Bytes buffer)
```

Comment:
This function does not support `record` being null, so it should take `record` as a reference.